lxd/backup.go

```diff
@@ -0,0 +1,297 @@
+	err = b.state.Cluster.ContainerBackupRename(b.Name(), newName)
+	if err != nil {
+		return err
+	}
+
+	// Rename the directories and files
+	err = b.container.Storage().ContainerBackupRename(*b, newName)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Delete removes a container backup.
+func (b *backup) Delete() error {
+	ourStart, err := b.container.StorageStart()
+	if err != nil {
+		return err
+	}
+	if ourStart {
+		defer b.container.StorageStop()
+	}
+
+	// Remove the database record
+	err = b.state.Cluster.ContainerBackupRemove(b.Name())
+	if err != nil {
+		return err
+	}
+
+	// Delete backup from storage
+	err = b.container.Storage().ContainerBackupDelete(b.Name())
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Dump dumps the container including its snapshots.
+func (b *backup) Dump() ([]byte, error) {
+	ourStart, err := b.container.StorageStart()
+	if err != nil {
+		return nil, err
+	}
+	if ourStart {
+		defer b.container.StorageStop()
+	}
+
+	data, err := b.container.Storage().ContainerBackupDump(*b)
+	if err != nil {
+		return nil, err
+	}
+
+	return data, nil
+}
+
+func (b *backup) Render() interface{} {
+	return &api.ContainerBackup{
+		Name:             b.name,
+		CreationDate:     b.creationDate,
+		ExpiryDate:       b.expiryDate,
+		ContainerOnly:    b.containerOnly,
+		OptimizedStorage: b.optimizedStorage,
+	}
+}
+
+func (b *backup) Id() int {
+	return b.id
+}
+
+func (b *backup) Name() string {
+	return b.name
+}
+
+func (b *backup) CreationDate() time.Time {
+	return b.creationDate
+}
+
+func (b *backup) ExpiryDate() time.Time {
+	return b.expiryDate
+}
+
+func (b *backup) ContainerOnly() bool {
+	return b.containerOnly
+}
+
+func (b *backup) OptimizedStorage() bool {
+	return b.optimizedStorage
+}
+
+func getBackupInfo(r io.Reader) (*backupInfo, error) {
+	var buf bytes.Buffer
+	err := shared.RunCommandWithFds(r, &buf, "unxz", "-")
+	if err != nil {
+		return nil, err
+	}
+
+	result := backupInfo{}
+	hasBinaryFormat := false
+	hasIndexFile := false
+	tr := tar.NewReader(&buf)
+	for {
+		hdr, err := tr.Next()
+		if err == io.EOF {
+			break // End of archive
+		}
+		if err != nil {
+			return nil, err
+		}
+
+		if hdr.Name == "backup/index.yaml" {
+			err = yaml.NewDecoder(tr).Decode(&result)
+			if err != nil {
+				return nil, err
+			}
+
+			hasIndexFile = true
+		}
+
+		if hdr.Name == "backup/container.bin" {
+			hasBinaryFormat = true
+		}
+	}
+
+	if !hasIndexFile {
+		return nil, ErrMissingIndexFile
+	}
+
+	result.HasBinaryFormat = hasBinaryFormat
+	return &result, nil
+}
+
+// fixBackupStoragePool changes the pool information in the backup.yaml. This
+// is done only if the provided pool doesn't exist. In this case, the pool of
+// the default profile will be used.
+func fixBackupStoragePool(c *db.Cluster, b backupInfo) error {
+	// Get the default profile
+	_, profile, err := c.ProfileGet("default")
+	if err != nil {
+		return err
+	}
+
+	_, v, err := shared.GetRootDiskDevice(profile.Devices)
```

Comment:
What if there's no root disk in the devices file? Won't you fail the `StoragePoolGet()` call then? Or did we decide that this is ok?

Reply:
Unless I'm missing something, we'd only get here if the storage pool listed in backup.yaml doesn't exist, we're then assuming the default profile exists and that it has a storage pool. If either of those isn't true, we fail, which seems reasonable.